### DISCUSSION TITLE
사용자 (검색) 기록 데이터베이스화 의사결정

### DISCUSSION THREAD

**Discussion by koreanddinghwan** [manito42/backend#73] [Ideas]
## 도입

사용자 기록 데이터화하는 모듈 설계중입니다.  
이에따라 스키마가 변경되어야하는데, 
현재로서는 기록(History)에 대해 '검색'이라는 행위에 대해서, 특히 mentorProfile에 대해서만 남기려고 합니다.  

<br>

## 현재 고민중인 부분

추후에 다른 기록, 즉
-  `유저의 행위를 데이터베이스화하고 이를 분석해 UX를 향상시킨다.`
라는 큰 그림으로 설계하면 이후에는 mentorProfile 로그 이외에 `다양한 로그 추가의 가능성`이 있다고 생각합니다.  
(ex, 유저의 페이지 이탈 이벤트 로그 등)

<br>

mentorProfileSearch라는 별도의 테이블 생성

유저의 행위 하나하나를 '테이블화' 하는것이 효율적인가?에 대해서는 저는 부정적으로 보는게,
나중에 로그로 만들어야할 형태의 기록마다 테이블화 하면 
1. 스키마 복잡성
2. 불필요한 인덱스 생성 등 문제가 발생할 것 같습니다.

<br>

## 의사결정

1. 사용자 검색기록 로그를 만들것인가
2. 로그를 만든다면, 현재 데이터베이스 그대로 사용 or Document Database 추가 사용?
     2.1 현재 데이터베이스 그대로 사용? => (로그 형태별로 테이블화 OR 로그를 [Json화해서 칼럼에 저장](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#json) )
     2.2 Doc DB는 무엇을 사용?(MongoDB, DynamoDB 등)

@JuneParkCode

**Comment by JuneParkCode** [comment]
1. 사용자 검색 로그 -> 일단은 GA 로 하고 있는게 좋지 않을까 생각을 해봅니다. 별도로 데이터 수집을 저장하고자 한다면, logger 를 이용해 search 쪽만 별도의 파일로 구성해보는건 어떨지? (추후에 데이터 옮길 수 있도록)

2. 이제 직접 관리 대상이 된다면, 데이터는 rdb 보다는 document 방식으로 처리하는게 나아보입니다. 말씀대로 테이블 하나 하나 두고 관리하기에는 부담이 있어 보입니다.

**Comment by koreanddinghwan** [comment]
> 1. 사용자 검색 로그 -> 일단은 GA 로 하고 있는게 좋지 않을까 생각을 해봅니다. 별도로 데이터 수집을 저장하고자 한다면, logger 를 이용해 search 쪽만 별도의 파일로 구성해보는건 어떨지? (추후에 데이터 옮길 수 있도록)

1. GA에서 사용자 검색 로그남기는게 프론트엔드 단에서 가능한지 모르겠네요
2. 파일 포맷이 database에서 지원하는 포맷과 다르면 파싱하기 상당히 귀찮아지게됩니다. 

<br><br>

> 2. 이제 직접 관리 대상이 된다면, 데이터는 rdb 보다는 document 방식으로 처리하는게 나아보입니다. 말씀대로 테이블 하나 하나 두고 관리하기에는 부담이 있어 보입니다.

- rdb내에서 document db특징을 그대로 사용할 수 있습니다. [JSON column](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#json)
- 이거 사용하면 테이블 하나하나 관리하는게 아닌, History라는 테이블 하나로만 관리가 가능해집니다.

**Comment by JuneParkCode** [comment]
1. 내부 검색에 대해서 별도로 들고오는거 가능한 것 같아요. 쇼핑몰 내부 검색이런거...
https://support.google.com/analytics/answer/1012264?hl=ko#zippy=%2C%EC%9D%B4-%EB%8F%84%EC%9B%80%EB%A7%90%EC%97%90%EC%84%9C%EB%8A%94-%EB%8B%A4%EC%9D%8C-%EC%82%AC%ED%95%AD%EC%9D%84-%EB%8B%A4%EB%A3%B9%EB%8B%88%EB%8B%A4

2. 이쪽은 제가 잘 모르는 부분이라 지금 섣불리 결정을하기가 어렵네요.

**Comment by koreanddinghwan** [comment]
> 1. 내부 검색에 대해서 별도로 들고오는거 가능한 것 같아요. 쇼핑몰 내부 검색이런거...
>    https://support.google.com/analytics/answer/1012264?hl=ko#zippy=%2C%EC%9D%B4-%EB%8F%84%EC%9B%80%EB%A7%90%EC%97%90%EC%84%9C%EB%8A%94-%EB%8B%A4%EC%9D%8C-%EC%82%AC%ED%95%AD%EC%9D%84-%EB%8B%A4%EB%A3%B9%EB%8B%88%EB%8B%A4
> 2. 이쪽은 제가 잘 모르는 부분이라 지금 섣불리 결정을하기가 어렵네요.

GA를 넣고 돌려봐야 알겠네요...

**Comment by koreanddinghwan** [comment]
일단, GA를 적용해보고 결정해야하는 사항 같습니다. ISSUE는 닫고, discussion은 열어놓겠습니다.
